sw/device/tests/sim_dv/all_escalation_resets_test.c

```diff
@@ -0,0 +1,889 @@
+__attribute__((section(".non_volatile_scratch")))
+const volatile uint32_t nv_reset_counter = UINT32_MAX;
+
+// This is a strike counter to keep track of progress.
+__attribute__((section(".non_volatile_scratch")))
+const volatile uint32_t nv_interrupt_counter = UINT32_MAX;
+
+/**
+ * Program the alert handler to escalate on alerts upto phase 2 (i.e. reset).
+ * Also program the aon timer with:
+ * - bark after escalation starts, so the interrupt is suppressed by escalation,
+ * - bite after escalation reset, so we should not get timer reset.
+ */
+enum {
+  kWdogBarkMicros = 10 * 100,         //   1 ms
+  kWdogBiteMicros = 15 * 100,         // 1.5 ms
+  kEscalationPhase0Micros = 2 * 100,  // 200 us
+  kEscalationPhase1Micros = 100,      // 100 us
+  kEscalationPhase2Micros = 100,      // 100 us
+  kMaxResets = 2,
+  kMaxInterrupts = 40,
+};
+
+static_assert(kWdogBarkMicros < kWdogBiteMicros &&
+                  kWdogBarkMicros > kEscalationPhase0Micros,
+              "The wdog bite shall happen only if escalation reset fails.");
+
+/**
+ * Objects to access the peripherals used in this test via dif API.
+ */
+static const uint32_t kPlicTarget = kTopEarlgreyPlicTargetIbex0;
+static dif_aes_t aes;
+static dif_alert_handler_t alert_handler;
+static dif_aon_timer_t aon_timer;
+static dif_clkmgr_t clkmgr;
+static dif_flash_ctrl_state_t flash_ctrl_state;
+static dif_kmac_t kmac;
+static dif_lc_ctrl_t lc_ctrl;
+static dif_otp_ctrl_t otp_ctrl;
+static dif_pwrmgr_t pwrmgr;
+static dif_rom_ctrl_t rom_ctrl;
+static dif_rstmgr_t rstmgr;
+static dif_rv_core_ibex_t rv_core_ibex;
+static dif_rv_plic_t plic;
+static dif_sram_ctrl_t sram_ctrl_main;
+static dif_sram_ctrl_t sram_ctrl_ret;
+
+static const char *we_check = "prim_reg_we_check";
+
+volatile bool interrupt_seen = false;
+
+void save_fault_checker(fault_checker_t *fault_checker) {
+  uint32_t function_addr = (uint32_t)(fault_checker->function);
+  uint32_t ip_inst_addr = (uint32_t)(fault_checker->ip_inst);
+  uint32_t type_addr = (uint32_t)(fault_checker->type);
+  CHECK(flash_ctrl_testutils_write(
+      &flash_ctrl_state, (uint32_t)(&nv_fault_checker[0]), 0, &function_addr,
+      kDifFlashCtrlPartitionTypeData, 1));
+  CHECK(flash_ctrl_testutils_write(
+      &flash_ctrl_state, (uint32_t)(&nv_fault_checker[1]), 0, &ip_inst_addr,
+      kDifFlashCtrlPartitionTypeData, 1));
+  CHECK(flash_ctrl_testutils_write(
+      &flash_ctrl_state, (uint32_t)(&nv_fault_checker[2]), 0, &type_addr,
+      kDifFlashCtrlPartitionTypeData, 1));
+}
+
+void restore_fault_checker(fault_checker_t *fault_checker) {
+  fault_checker->function =
+      (FaultCheckerFunction)(abs_mmio_read32((uint32_t)(nv_fault_checker + 0)));
+  fault_checker->ip_inst =
+      (char *)(abs_mmio_read32((uint32_t)(nv_fault_checker + 1)));
+  fault_checker->type =
+      (char *)(abs_mmio_read32((uint32_t)(nv_fault_checker + 2)));
+}
+
+static const char *no_name = "unidentified";
+
+// It would be handy to generate these.
```

Comment:
you could template and autogenerate this test if you wanted like the alert_test.c: https://cs.opensource.google/opentitan/opentitan/+/master:util/topgen/templates/alert_test.c.tpl?q=alert_test

But its ok have you have it too. Just throwing the option out there in case you were unaware this was possible.

Reply:
Part of the test can be templated, but the various <ip>_fault_checker cannot since the CSR bit set for these alerts is not standard. I could place them in a separate file... oh well, not on this PR.